pinot-query-planner/src/main/java/org/apache/pinot/calcite/rel/rules/PinotImplicitTableHintRule.java

```diff
@@ -0,0 +1,283 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.calcite.rel.rules;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.Nullable;
+import org.apache.calcite.plan.RelOptRuleCall;
+import org.apache.calcite.plan.RelRule;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.hint.RelHint;
+import org.apache.calcite.rel.logical.LogicalTableScan;
+import org.apache.pinot.calcite.rel.hint.PinotHintOptions;
+import org.apache.pinot.common.utils.DatabaseUtils;
+import org.apache.pinot.core.routing.TablePartitionInfo;
+import org.apache.pinot.spi.config.table.TableType;
+import org.apache.pinot.spi.utils.builder.TableNameBuilder;
+import org.immutables.value.Value;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@Value.Enclosing
+public class PinotImplicitTableHintRule extends RelRule<RelRule.Config> {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(PinotImplicitTableHintRule.class);
+  private final PartitionTableFinder _partitionTableFinder;
+
+  private PinotImplicitTableHintRule(Config config) {
+    super(config);
+    _partitionTableFinder = config.getTablePartitionTableFinder();
+  }
+
+  public static PinotImplicitTableHintRule withPartitionTableFinder(PartitionTableFinder tablePartitionTableFinder) {
+    return new PinotImplicitTableHintRule(ImmutablePinotImplicitTableHintRule.Config.builder()
+        .operandSupplier(b0 -> b0.operand(LogicalTableScan.class).anyInputs())
+        .tablePartitionTableFinder(tablePartitionTableFinder)
+        .build()
+    );
+  }
+
+  @Override
+  public boolean matches(RelOptRuleCall call) {
+    LogicalTableScan tableScan = call.rel(0);
+
+    RelHint explicitHint = getTableOptionHint(tableScan);
+
+    if (explicitHint == null) {
+      return true;
+    }
+    // we don't want to apply this rule if the explicit hint is complete
+    Map<String, String> kvOptions = explicitHint.kvOptions;
+    return kvOptions.containsKey(PinotHintOptions.TableHintOptions.PARTITION_KEY)
+        && kvOptions.containsKey(PinotHintOptions.TableHintOptions.PARTITION_FUNCTION)
+        && kvOptions.containsKey(PinotHintOptions.TableHintOptions.PARTITION_SIZE)
+        && kvOptions.containsKey(PinotHintOptions.TableHintOptions.PARTITION_PARALLELISM);
```

Comment:
Partition parallelism shouldn't be set by this rule. We may remove all logic around it

Reply:
Removed from here